dids/did_dht.go

```diff
@@ -0,0 +1,316 @@
+package dids
+
+import (
+	"encoding/base64"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"golang.org/x/net/dns/dnsmessage"
+
+	"github.com/tbd54566975/web5-go/crypto/dsa"
+	"github.com/tv42/zbase32"
+)
+
+var txtEntityNames = map[string]struct{}{
+	"vm":   {},
+	"auth": {},
+	"asm":  {},
+	"agm":  {},
+	"inv":  {},
+	"del":  {},
+}
+
+var relationshipDNStoDID = map[string]string{
+	"auth": "authentication",
+	"asm":  "assertionMethod",
+	"agm":  "keyAgreement",
+	"inv":  "capabilityInvocation",
+	"del":  "capabilityDelegation",
+}
+
+var keyTypes = map[string]string{
+	"0": dsa.AlgorithmIDED25519,
+	"1": dsa.AlgorithmIDSECP256K1,
+	"2": dsa.AlgorithmIDSECP256K1,
+}
+
+// dhtDIDRecord is used to structure the DNS representation of a DID
+type dhtDIDRecord struct {
+	rootRecord string
+	records    map[string]string
+}
+
+func (rec *dhtDIDRecord) DIDDocument(didURI string) *Document {
+	relationshipMap := parseVerificationRelationships(rec.rootRecord)
+
+	// Now we have a did in a dns record. yay
+	document := &Document{
+		ID: didURI,
+	}
+
+	// Now create the did document
+	for name, data := range rec.records {
+
+		switch {
+		case strings.HasPrefix(name, "_k"):
+			vMethod, err := UnmarshalVerificationMethod(data)
+			if err != nil {
+				// TODO handle error
+			}
+
+			// extracting kN from _kN._did
+			entryId := strings.Split(name, ".")[0][1:]
+			relationships, ok := relationshipMap[entryId]
+
+			if !ok {
+				// no relationships
+				continue
+			}
+
+			opts := []string{}
+			for _, r := range relationships {
+				if o, ok := relationshipDNStoDID[r]; ok {
+					opts = append(opts, o)
+				}
+			}
+
+			document.AddVerificationMethod(
+				*vMethod,
+				Purposes(opts...),
+			)
+		case strings.HasPrefix(name, "_s"):
+			s := UnmarshalService(data)
+			document.AddService(s)
+		case strings.HasPrefix(name, "_cnt"):
+			// TODO add controller https://did-dht.com/#controller
+			// optional field
+			// comma-separated list of controller DID identifiers
+			document.Controller = strings.Split(data, ",")
+		case strings.HasPrefix(name, "_aka"):
+			// TODO add aka https://did-dht.com/#also-known-as
+			document.AlsoKnownAs = strings.Split(data, ",")
+		default:
+		}
+	}
+
+	return document
+}
+
+// ResolveDIDDHT resolves a DID using the DHT method
+func ResolveDIDDHT(uri, relay string, client *http.Client) (ResolutionResult, error) {
```

Comment:
Why parameterize the http client?

Reply:
so... https://github.com/TBD54566975/web5-go/pull/14 changes this, but, generally it's a good idea to not make the http client implicit because a consumer of this method might want to set up their own configuration of the cli (think timeouts, possible backoff, breakers, tracing headers maybe, etc) we can default to one for them, but giving the option to configure their one is usually a good idea